### PR TITLE
[VL] Optimize Delta DV applyDeletionFilter with iterator-based bulk lookup

### PR DESCRIPTION
**Pull request by iemejia**
## What changes are proposed in this pull request?

Replace per-row `contains()` calls with an iterator-based scan in `DeltaDeletionVectorReader::applyDeletionFilter()`.

**Before:** O(batch_size) calls to `Roaring64Map::contains()`, each performing a `std::map` lookup (O(log N) on the high-32-bit buckets) plus a Roaring container check per row position.

**After:** O(deletions_in_range) iterator advances using `move_equalorlarger()` to seek directly to the first deleted row in the batch range, then sequential `++iterator` advances (O(1) amortized within a Roaring container) until the range end.

### Benchmark results (1M row file, batches of 4096, scanning full file)

| Deletion % | Baseline (ms) | Optimized (ms) | Speedup |
|---|---|---|---|
| **1% (sparse)** | **9.90** | **0.050** | **198x** |
| 10% (moderate) | 4.06 | 0.398 | 10.2x |
| 50% (dense) | 3.82 | 1.99 | 1.9x |
| 90% (very dense) | 4.37 | 3.64 | 1.2x |

For the common case of sparse deletions (1%, typical for Delta MERGE/UPDATE), this is a **198x improvement**.

The baseline O(batch_size) approach spent constant time (~4ms per 1M rows) regardless of deletion density because it always probed every row. The optimized O(deletions) approach scales linearly with actual deletions.

## How was this patch tested?

- 21 C++ unit tests pass (12 existing + 9 new corner cases):
  - All rows deleted / no rows deleted
  - First and last row of batch deleted
  - Large 64-bit offsets (multi-bucket Roaring64Map)
  - Single-row batches (deleted / not deleted)
  - Dense/sparse mixed deletion patterns
  - Batch range entirely before / after all deletions
  - Partial overlap with deletion range
- New `BM_ApplyDeletionFilter` benchmark added to `DeltaBitmapBenchmark.cc` with 5 scenarios (sparse/moderate/dense/very-dense/large-batch).

## Was this patch authored or co-authored using generative AI tooling?

Generated-by: Claude claude-opus-4.6

### CODE REVIEW

File: cpp/velox/benchmarks/DeltaBitmapBenchmark.cc [hunk 1]
```diff
@@ -276,4 +280,78 @@ BENCHMARK_CAPTURE(
...
+      }
+    }
+    benchmark::DoNotOptimize(totalDeletedFound);
+  }
+
+  state.SetItemsProcessed(state.iterations() * totalFileRows);
+  state.counters["batch_size"] = benchmark::Counter(batchSize);
+  state.counters["deletion_pct"] = benchmark::Counter(deletionPercent);
+  state.counters["deleted_found"] = benchmark::Counter(totalDeletedFound);
+  state.counters["total_batches"] = benchmark::Counter(numBatches);
```
[the remark this answers is not in the document]
Comment (iemejia) [reply cpp/velox/benchmarks/DeltaBitmapBenchmark.cc:333]: Good catch on both points.

**Tail rows:** Fixed — `SetItemsProcessed` now uses `numBatches * batchSize` (the rows actually processed) instead of `totalFileRows`. For the default batch size of 4096, the tail is 576 rows out of 1M (0.06%), so the benchmark results are unaffected, but the counter is now accurate.

**Padding bits in popcount:** This is actually safe — `applyDeletionFilter` calls `memset(rawBitmap, 0, nbytes(size))` at the top, which zeros the entire bitmap including padding bits in the last word. Added an inline comment clarifying this.

File: cpp/velox/compute/delta/DeltaDeletionVectorReader.cpp [hunk 2]
```diff
@@ -183,15 +183,29 @@ void DeltaDeletionVectorReader::applyDeletionFilter(uint64_t baseReadOffset, uin
   auto* rawBitmap = deleteBitmap->asMutable<uint64_t>();
   std::memset(rawBitmap, 0, bits::nbytes(size));
 
+  // Use an iterator-based approach instead of per-row contains() lookups.
+  // This is O(deletions_in_range) rather than O(batch_size), which is
+  // significantly faster when deletions are sparse relative to batch size.
+  const uint64_t rangeEnd = baseReadOffset + size;
```
[the remark this answers is not in the document]
Comment (iemejia) [reply cpp/velox/compute/delta/DeltaDeletionVectorReader.cpp]: Fixed — added saturating arithmetic so `rangeEnd` clamps to `UINT64_MAX` instead of wrapping:

```cpp
const uint64_t rangeEnd = (size <= UINT64_MAX - baseReadOffset)
    ? baseReadOffset + size : UINT64_MAX;
```

Also added a unit test (`ApplyDeletionFilterOverflowProtection`) that exercises `baseReadOffset = UINT64_MAX - 50` with `size = 100` to verify the guard works correctly.

File: cpp/velox/benchmarks/DeltaBitmapBenchmark.cc [hunk 3]
```diff
@@ -276,4 +280,69 @@ BENCHMARK_CAPTURE(
...
+  }
+  const auto payload = bitmap.serializeToString(true);
+
+  // Load the DV reader.
+  DeltaDeletionVectorReader reader;
+  reader.loadSerializedDeletionVector(std::string_view(payload.data(), payload.size()));
+
+  // Allocate the output bitmap buffer.
+  memory::MemoryManager::testingSetInstance(memory::MemoryManager::Options{});
+  auto pool = memory::memoryManager()->addLeafPool();
```
[the remark this answers is not in the document]
Comment (iemejia) [reply cpp/velox/benchmarks/DeltaBitmapBenchmark.cc:305]: Good catch — moved `testingSetInstance` out of the benchmark body into a custom `main()` that initializes once before running benchmarks. This mirrors the pattern in `GenericBenchmark.cc:642`.

File: cpp/velox/compute/delta/tests/DeltaDeletionVectorReaderTest.cpp [hunk 4]
```diff
@@ -217,3 +217,186 @@ TEST_F(DeltaDeletionVectorReaderTest, BatchFilteringPartialOverlap) {
...
+  auto deleteBitmap = AlignedBuffer::allocate<uint64_t>(bits::nwords(100), pool_.get());
+  reader.applyDeletionFilter(nearMax, 100, deleteBitmap);
+
+  auto* rawBitmap = deleteBitmap->as<uint64_t>();
+  // Rows at relative positions 10 and 20 should be marked as deleted.
+  EXPECT_TRUE(bits::isBitSet(rawBitmap, 10));
+  EXPECT_TRUE(bits::isBitSet(rawBitmap, 20));
+  EXPECT_FALSE(bits::isBitSet(rawBitmap, 0));
+  EXPECT_FALSE(bits::isBitSet(rawBitmap, 30));
+}
```
[the remark this answers is not in the document]
Comment (iemejia) [reply cpp/velox/compute/delta/tests/DeltaDeletionVectorReaderTest.cpp:403]: Good catch. Fixed in c56673720 — the test now uses low-value bitmap entries (`{10, 20, 30}`) that are valid under `addSafe()`'s `kMaxRepresentableValue` limit, and exercises the overflow guard purely through a high `baseReadOffset` (`UINT64_MAX - 50` with `size = 100`). This verifies the saturating arithmetic in `rangeEnd` without needing to bypass the serialization validation.

Constructing a raw Roaring64Map payload with near-UINT64_MAX positions would let us test deletions actually present in the overflow range, but Delta's spec caps row indices well below that, so such a DV would be invalid in practice.

**Review by zhztheplayer — APPROVED** [review]

**Comment by iemejia** [comment]
Thanks @zhztheplayer !

**Comment by felipepessoto** [comment]
Contributes to #11901
